core/src/main/java/io/grpc/transport/MessageDeframer.java

```diff
@@ -321,31 +483,37 @@ private void processHeader() {
     compressedFlag = (type & COMPRESSED_FLAG_MASK) != 0;
 
     // Update the required length to include the length of the frame.
-    requiredLength = nextFrame.readInt();
+    requiredLength = header.readInt();
 
     // Continue reading the frame body.
     state = State.BODY;
   }
 
   /**
-   * Processes the body of the GRPC compression frame. A single compression frame may contain
-   * several GRPC messages within it.
+   * Processes the body of the GRPC frame.
    */
-  private void processBody() {
-    InputStream stream = compressedFlag ? getCompressedBody() : getUncompressedBody();
-    nextFrame = null;
-    listener.messageRead(stream);
+  private void processBody(CompositeReadableBuffer body, ArrayDeque<Frame> queue) {
+    InputStream stream = compressedFlag ? getCompressedBody(body) : getUncompressedBody(body);
+    // Add the body to the queue and consume a flow-control token
+    transportCount.getAndDecrement();
+    Frame last = queue.peekLast();
+    if (last != null && last.stream == null) {
+      // Can re-use the tail entry
+      last.stream = stream;
```

Comment:
Are we sure this is safe? We need to be sure that last is not `END_OF_STREAM` (it's static).